include/gatekeeper_cps.h

```diff
@@ -92,6 +92,14 @@ struct cps_config {
 	unsigned int      front_kni_index;
 	unsigned int      back_kni_index;
 
+	/*
+	 * MTUs for the KNIs; used with routing daemons.
+	 * Read from the KNI using Netlink to initialize,
+	 * and then updated on calls to kni_change_mtu().
+	 */
+	unsigned int      front_kni_mtu;
+	unsigned int      back_kni_mtu;
```

Comment:
OK, so I'm hearing two things.

1. Change `kni_change_mtu()` to disable being able to change the MTU. We therefore also don't need to update the KNI MTU.

2. Still, we should calculate the MTU of the KNI at boot. Is there a reason why we cannot just ask the kernel for the MTU using Netlink as is done in this patch? Then we can compare it to the physical interface's MTU. Note that this is sometimes different -- our NICs support a higher MTU (we use 2048 on the back interface) but the KNIs appear to only support up to 1500.

If we can't ask for the MTU from the kernel, I'm not sure how to calculate the size of the link layer header since that is per-packet and not necessarily associated with the interface.

Reply:
1. Correct.

2. If we can't change the size of the MTU of a KNI interface at boot, what will happen when the back interface receives a packet of 2048 and we need to handle it to the kni interface whose MTU is 1500?

I don't understand why we need to ask the MTU of the physical interface to the kernel since this information is available at `gatekeeper_if.mtu` for front and back interfaces in `net`. And the length of the link layer header is available at `gatekeeper_if.l2_len_out`.

Reply:
According to [this page](https://doc.dpdk.org/guides/prog_guide/kernel_nic_interface.html) (see `struct rte_kni_conf`) we can set the MTU size.

Reply:
To answer your question about what happens when there is a mismatch in MTU, we have this comment:

https://github.com/AltraMayor/gatekeeper/blob/master/include/gatekeeper_net.h#L156-L158

I actually don't know if the packet would be dropped or if it would be truncated, but perhaps that is worth checking once we more explicitly track the MTU of the KNI.

Sorry, I was asking whether we should request the MTU of the KNI from the kernel. Yes, we can always get the MTU of the Gatekeeper interfaces from our configuration, but to get the MTU of the KNI this patch requests it from the kernel. That way, when the daemon asks for link information we report the minimum out of {Gatekeeper interface MTU, KNI MTU}.

Reply:
The MTU of the KNI appears to have a limit of 1500 from the virtual device that the KNI library creates, so we can specify it in `struct rte_kni_conf` but only up to 1500. It might be worth specifying it as 1500 so that we can assume that this is the case and not need to ask the kernel.

Reply:
Yes. Let's set the MTU explicitly. We can keep the code to ask the kernel the MTU to make sure that the MTU we asked was respected.

Could you open an issue to test and (if needed) solve the issue that arrises when the MTU of the KNI interface would need to be larger than 1500? This might prevent the BGP speaker of working. This issue should go the First Deployment milestone.

Reply:
I had it wrong. If you don't set the `mtu` field in `struct rte_kni_conf` then you get a default value of 1500 and any attempts to change it are unsuccessful (hence me thinking 1500 was a hard limit). But setting that field indeed allows you to set the MTU much higher (I tried 64000 and it worked).

So I will add a patch to disable changing the MTU, add a patch to set the KNI MTU to whatever is in the configuration for the physical interface's MTU and verify it with the kernel, and then simplify the current patch to just report whatever the configuration has as the MTU.

Reply:
That is great news!

Two observations:

1. The MTU of the KNI should be the physical MTU - physical Link layer header + KNI link layer header. And verify during boot that physical Link layer header >= KNI link layer header.

2. Given that we can set the KNI MTU for whatever we need, we no longer to need to double check the MTU with the kernel.

Reply:
Setting the KNI's MTU to be slightly smaller to account for the case where we need to use a larger L2 header for the physical makes sense. Up to this point, we have assumed that the KNI does not have a VLAN ID, and insert one if the physical interface does:

https://github.com/AltraMayor/gatekeeper/blob/master/cps/main.c#L430-L456

If we keep this assumption, then when the physical interface has a VLAN ID we can just make the MTU of the KNI to be the `physical MTU - sizeof(struct vlan_hdr.vlan_tci)`.

If we remove this assumption, then I think we will need to ask the kernel for link information about the interface and parse whether the response has a `IFLA_VLAN_ID` attribute.

Reply:
We should keep the assumption that the KNI does not have a VLAN ID. This keeps the code as it is, which I think is its simplest form because it requires no change in any daemon using the KNI interfaces and the setup of the KNI interfaces for the Linux kernel is straightforward.